File: erl/hanoi.erl
%% The Towers Of Hanoi
%% Erlang
%% Copyright (C) 2002 Amit Singh. All Rights Reserved.
%% http://hanoi.kernelthread.com
%%
%% Last tested under Erlang (BEAM) emulator version 5.0.2.4
%%

-module(hanoi).
-export([hanoi/1]).
-author('Amit Singh').

m(F,T) -> io:format("~w~s~w~n", [F, " --> ", T]).

h(1,F,U,T) -> m(F, T);
h(N,F,U,T) when N > 0 ->
    h(N - 1, F, T, U),
    m(F, T),
    h(N - 1, U, F, T).

hanoi(0) -> io:format("number of disks cannot be 0~n");
hanoi(N) when N < 0 -> io:format("number of disks must be > 0~n");
hanoi(N) when N > 10 -> io:format("number of disks must be < 10~n");
hanoi(N) -> h(N, 1, 2, 3).

File: hs/hanoi.hs
m :: (Num a) => ([a],[a],[a]) -> ([a],[a],[a])
m (h:f, u, t) = (f, u, h:t)

a :: (Num a) => ([a],[a],[a]) -> ([a],[a],[a])
a (f, t, u) = (f, u, t)

b :: (Num a) => ([a],[a],[a]) -> ([a],[a],[a])
b (f, u, t) = (u, f, t)

c :: (Num a) => ([a],[a],[a]) -> ([a],[a],[a])
c (u, f, t) = (f, u, t)

h :: (Num a) => a -> ([a],[a],[a]) -> ([a],[a],[a])
h 1 (f, u, t) = m (f, u, t)
h n (f, u, t) = c (h (n-1) (b (m (a (h (n-1) (f, t, u)) ))))

hanoi :: (Num a) => a -> [a] -> ([a],[a],[a])
hanoi n l = h n (l, [], [])

main = print (hanoi 23 [1..23])
